### PR TITLE
Core/adding table reading capability bc loads V2

### PR DESCRIPTION
**Pull request by AlejandroCornejo**
In this PR I'm adding to the core an implemenation carried out initially by @ipouplana regarding the application of Boundary Conditions and loads whose value is defined according to a Table.
The json format is:
```
{
    "python_module": "apply_vector_constraint_table_process",
            "kratos_module": "KratosMultiphysics",

            "Parameters":    {
                "model_part_name": "Structure.DISPLACEMENT_Displacement_Auto2",
                "variable_name":   "DISPLACEMENT",
                "time_variable_name" :  "TIME",
                "active":          [true,true,true],
                "is_fixed":        [true,true,true],
                "value":           [0.0,0.0,0.0],
                "table":           [1,0,0]
            }
        }
```
Where the key "table" defines the table that is goberning that direction.

Inside the mdpa ons must include the table as:
```
Begin Table 1 TIME DISPLACEMENT_X
  0.0 0.1
  1.0 0.2
  10 0.2
End Table
```

and including the defined table inside the SubModelPart:

```
Begin SubModelPart DISPLACEMENT_Displacement_Auto2
    Begin SubModelPartTables
	1
	End SubModelPartTables
	Begin SubModelPartNodes
            1
            2
            6
           11
           17
           25
           35
           47
           61
    End SubModelPartNodes
    Begin SubModelPartElements
    End SubModelPartElements
    Begin SubModelPartConditions
    End SubModelPartConditions
End SubModelPart
```

Ive done this PR because the Git has mixed two branches and I do not know why...
I've solved some requirements but it's still work to do.
Sorry for the inconvinience

### CODE REVIEW

File: kratos/processes/apply_component_table_process.cpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    
+    unsigned int TableId = rParameters["table"].GetInt();
+    mpTable = rModelPart.pGetTable(TableId);
+
+    KRATOS_CATCH("");
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+/// this function is designed for being called at the beginning of the computations
```
Comment (loumalouomega) [inline kratos/processes/apply_component_table_process.cpp]: Description should be on headers

File: kratos/processes/apply_component_table_process.h [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+
+protected:
+
+    /// Copy constructor.
+    ApplyComponentTableProcess(ApplyComponentTableProcess const& rOther);
+
+    /// Member Variables
+    ModelPart& mrModelPart;
```
Comment (loumalouomega) [inline kratos/processes/apply_component_table_process.h]: Add some description to the member variables

File: kratos/processes/apply_double_table_process.h [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+// Project includes
+#include "processes/apply_component_table_process.h"
+
+namespace Kratos
+{
+
+/// The base class for all processes in Kratos.
+/** This function applies a value according to a table defined in the mdpa
+*/
+class KRATOS_API(KRATOS_CORE) ApplyDoubleTableProcess : public ApplyComponentTableProcess
```
Comment (loumalouomega) [inline kratos/processes/apply_double_table_process.h]: Please, use templates to avoid code duplication, this is a blocker

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Please use templates to reduce code duplication

**Comment by philbucher** [comment]
I like this feature!

Can you please wait for my Feedback?
I am back in the office on Monday 
Thanks!

**Comment by AlejandroCornejo** [comment]
I've unified the classes into one templated. The thing is that now it doesn't compile IDK why... I'v created a template like:


```
template<class TVariableType>
class KRATOS_API(KRATOS_CORE) ApplyComponentTableProcess 
    : public Process
```


```
template<>
void ApplyDoubleTableProcess<DoubleVariableType>::ExecuteInitializeSolutionStep()
{
...
}

template<>
void ApplyComponentTableProcess<ComponentVariableType>::ExecuteInitializeSolutionStep()
{
...
}
```


with the typedef's

```
typedef VariableComponent<VectorComponentAdaptor<array_1d<double,3>>>  ComponentVariableType;
typedef Variable<double>                                                  DoubleVariableType;
```


So the template is the type of variable (double or components)... Am I doing any wrong?

Thank you

File: kratos/processes/apply_component_table_process.cpp (hunk 1 above)
Comment (AlejandroCornejo) [reply kratos/processes/apply_component_table_process.cpp]: done

File: kratos/processes/apply_component_table_process.h (hunk 2 above)
Comment (AlejandroCornejo) [reply kratos/processes/apply_component_table_process.h]: done

File: kratos/processes/apply_double_table_process.h (hunk 3 above)
Comment (AlejandroCornejo) [reply kratos/processes/apply_double_table_process.h]: Now it's only one templated class

**Comment by AlejandroCornejo** [comment]
> I like this feature!
> 
> Can you please wait for my Feedback?
> I am back in the office on Monday
> Thanks!

Sure! have a look whenever you can and I'll review.
Thank you all

**Comment by loumalouomega** [comment]
@AlejandroCornejo is this ready for the final review?

**Comment by AlejandroCornejo** [comment]
Now it compiles and it works, i think so thank you :)

**Review by loumalouomega — COMMENTED** [review]
Huum, I need to think, you are using explicit specializations, I think it can be done in a simpler way

**Comment by loumalouomega** [comment]
@AlejandroCornejo I reduced the code duplication, it only misses a test

**Comment by loumalouomega** [comment]
If you add a test to the process I will approve

**Comment by AlejandroCornejo** [comment]
Ok! I'll implement the tests and wait for the comments from @philbucher

**Comment by AlejandroCornejo** [comment]
@loumalouomega do you mean a python or cpp test? thx

**Comment by loumalouomega** [comment]
Python test would be fine, as you prefer.

**Comment by philbucher** [comment]
I recommend python in the same way the other similar processes in the core are tested
=> kratos/tests/test_processes.py

**Comment by loumalouomega** [comment]
(and it doesn't increase compilation time ⌚)

**Review by ipouplana — COMMENTED** [review]
Thanks for the work @AlejandroCornejo ! I am sure that if you adapt these processes to the current standards they will be very useful.

**Review by philbucher — CHANGES_REQUESTED** [review]
I had a thorough look at this PR and have the following comments:
- The functionality you add is really helpful IMO, +1 for that
- I am quite convinced that the implementation can be simplified very much, bcs this process replicates basically the same functionality as the `AssignScalarVariableProcess`, except how the value is assigned
=> therefore I suggest instead of adding many files to slightly refactor the `AssignScalarVariableProcess` to make derivation possible. I made a draft here: https://github.com/KratosMultiphysics/Kratos/tree/core/table_proc_bcs

This should make your life also a lo easier since you only have to add / change how the value is computed.

For parameters I recommend to use what I recently [introduced in the solvers](https://github.com/KratosMultiphysics/Kratos/pull/4736) to not duplicate the settings

What do you think?

**Review by loumalouomega — DISMISSED** [review]

**Comment by AlejandroCornejo** [comment]
> I had a thorough look at this PR and have the following comments:
> 
> * The functionality you add is really helpful IMO, +1 for that
> * I am quite convinced that the implementation can be simplified very much, bcs this process replicates basically the same functionality as the `AssignScalarVariableProcess`, except how the value is assigned
>   => therefore I suggest instead of adding many files to slightly refactor the `AssignScalarVariableProcess` to make derivation possible. I made a draft here: https://github.com/KratosMultiphysics/Kratos/tree/core/table_proc_bcs
> 
> This should make your life also a lo easier since you only have to add / change how the value is computed.
> 
> For parameters I recommend to use what I recently [introduced in the solvers](https://github.com/KratosMultiphysics/Kratos/pull/4736) to not duplicate the settings
> 
> What do you think?

Hi! I've added all the requirements but currently I dont' have enough time to study those processes... :S what about if I do another PR modifying that? sorry but I have to move on with other stuff hehe

**Comment by philbucher** [comment]
Hm I am not sure ...
I would much more prefer to keep the core clean. Thing is that ppl are usually not happy if we break things, therefore I think it is better to do it right the first time.

I see you added a test, maybe I will just do it myself :)
From what I understand this is not urgent, give me a couple of days

**Comment by AlejandroCornejo** [comment]
Which is the state of this? @philbucher @loumalouomega ? :)

**Review by loumalouomega — APPROVED** [review]
OK from my side

**Comment by philbucher** [comment]
Sorry I am traveling last and this week, will try to do it next week

If you want you can also do it yourself with what I described 
It should be quite simple actually

**Review by rubenzorrilla — CHANGES_REQUESTED** [review]
Guys, at first glance there are some things that I'd like to comment. Besides I feel that we can generalize this feature a bit more (maybe I'm wrong). I'll try to have a look during the next days.

**Comment by AlejandroCornejo** [comment]
Honestly I think that we should reconsider to implement this in a formal way! What do you think @KratosMultiphysics/technical-committee ?

**Pull request closed without merging**